File: Iterator/Iterator.go
package main

import "fmt"

type Itertor interface {
	init(names []string,index int)
	hasNext() bool
	next() string
}

type NamesIteror struct {
	names []string
	index int
}

func(n *NamesIteror)init(names []string,index int){
	n.names=names
	n.index=index
}

func(n *NamesIteror)hasNext()bool{
	if(n.index<len(n.names)){
		return true;
	}
	return false;
}

func(n *NamesIteror)next() string{
	if(n.hasNext()){
		index:=n.index
		n.index++
		return n.names[index]
	}
	return "";
}


type Container interface {
	getItertor() Itertor
}

type NameRepository struct {
	names []string
	itertor Itertor
}

func(n *NameRepository)getItertor() Itertor{
	n.names=[]string{0:"Robert",1:"John",2:"Julie",3:"Lora"}
	n.itertor=new(NamesIteror)
	n.itertor.init(n.names,0)
	return n.itertor
}

func main(){
	//demo
	nameRepository:=new(NameRepository)
	nameR:=nameRepository.getItertor()
	for nr:=nameR;nr.hasNext();{
		fmt.Println("Name :",nr.next())
	}
}

File: Iterator/Iterator.php
<?php
/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/25
 * Time: 10:46
 */

interface Iterater{
    public function hasNext();
    public function next();
}

interface Container{
    public function getIterator();
}

class NameIterator implements Iterater{
    private $index=0;
    private $names;
    public function __construct(&$names)
    {
     $this->names=$names;
    }

    public function hasNext()
    {
        // TODO: Implement hasNext() method.
        if($this->index<count($this->names)){
            return true;
        }
        return false;
    }

    public function next()
    {
        // TODO: Implement next() method.
        if($this->hasNext()){
            return $this->names[$this->index++];
        }
        return null;
    }
}

class NameRepository implements Container{
    private $names=array("Robert","John","Julie","Lora");
    public function getIterator()
    {
        // TODO: Implement getIterator() method.
           return new NameIterator($this->names);
    }

}


//demo
$namesRepository=new NameRepository();
for($iter=$namesRepository->getIterator();$iter->hasNext();){
    $name=$iter->next();
    var_dump("Name :".$name);
}

